network/p2p/gossip/handler.go

```diff
@@ -112,8 +149,76 @@ func (h Handler[T]) AppRequest(_ context.Context, _ ids.NodeID, _ time.Time, req
 		Gossip: gossipBytes,
 	}
 
-	h.sentN.Add(float64(len(response.Gossip)))
-	h.sentBytes.Add(float64(responseSize))
+	h.pullGossipSentN.Add(float64(len(response.Gossip)))
+	h.pullGossipSentBytes.Add(float64(responseSize))
 
 	return proto.Marshal(response)
 }
+
+func (h Handler[T, U]) AppGossip(ctx context.Context, nodeID ids.NodeID, gossipBytes []byte) {
+	msg := &sdk.PushGossip{}
+	if err := proto.Unmarshal(gossipBytes, msg); err != nil {
+		h.log.Debug("failed to unmarshal gossip", zap.Error(err))
+		return
+	}
+
+	forward := make([][]byte, 0, len(msg.Gossip))
+	receivedBytes := 0
+	sentBytes := 0
+	for _, bytes := range msg.Gossip {
+		gossipable := U(new(T))
+		if err := gossipable.Unmarshal(bytes); err != nil {
+			h.log.Debug("failed to unmarshal gossip",
+				zap.Stringer("nodeID", nodeID),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		receivedBytes += len(bytes)
+
+		if _, ok := h.set.Get(gossipable.GetID()); ok {
+			continue
+		}
+
+		if err := h.set.Add(gossipable); err != nil {
+			h.log.Debug(
+				"failed to add gossip to the known set",
+				zap.Stringer("nodeID", nodeID),
+				zap.Stringer("id", gossipable.GetID()),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		// continue gossiping messages we have not seen to other peers
+		forward = append(forward, bytes)
+		sentBytes += len(bytes)
+	}
+
+	forwardMsg := &sdk.PushGossip{
+		Gossip: forward,
+	}
+
+	forwardMsgBytes, err := proto.Marshal(forwardMsg)
+	if err != nil {
+		h.log.Debug(
+			"failed to marshal forward gossip message",
+			zap.Error(err),
+		)
+		return
+	}
+
+	h.pushGossipReceivedN.Add(float64(len(msg.Gossip)))
+	h.pushGossipReceivedBytes.Add(float64(receivedBytes))
+	h.pushGossipSentN.Add(float64(len(forwardMsg.Gossip)))
+	h.pushGossipSentBytes.Add(float64(sentBytes))
+
+	if err := h.client.AppGossip(ctx, forwardMsgBytes); err != nil {
+		h.log.Debug(
+			"failed to forward gossip",
+			zap.Error(err),
+		)
+		return
+	}
```

Comment:
Should these errors be `Warn` (maybe even `Error`) rather than `Debug`? I don't think they should ever fail right?